server/service/src/number.rs

```diff
@@ -75,4 +59,133 @@ mod test {
+    async fn test_concurrent_next_number() {
+        let (_, _, connection_manager, _) = test_db::setup_all(
+            "test_concurrent_numbers",
+            MockDataInserts::none().names().stores(),
+        )
+        .await;
+
+        // Note: this test is disabled when running tests using in 'memory' sqlite.
+        // When running in memory sqlite uses a shared cache and returns an SQLITE_LOCKED response when two threads try to write using the shared cache concurrently
+        // https://sqlite.org/rescode.html#locked
+        // We are relying on busy_timeout handler to manage the SQLITE_BUSY response code in this test and there's no equivelant available for shared cache connections (SQLITE_LOCKED).
+        // If we were to use shared cache in production, we'd probably need to use a mutex (or similar) to protect the database connection.
+
+        /*
+            Test Scenario
+
+            Process A starts a transaction, and gets the next number, then waits before commiting the transaction
+            Concurrently Process B tries to get the next number
+            (Note: This test did fail with previous implementation of next number on postgres)
+        */
+
+        //Part 1: Concurrent up date (first row) e.g. this will require an insert and an update for one these processes...
+
+        let manager_a = connection_manager.clone();
+        let process_a = std::thread::spawn(move || {
+            let connection = manager_a.connection().unwrap();
+            let result: Result<i64, TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    let num = next_number(con, &NumberRowType::Stocktake, "store_a")?;
+                    std::thread::sleep(core::time::Duration::from_millis(TEST_SLEEP_TIME));
+                    Ok(num)
+                });
+            result.unwrap()
+        });
+
+        let manager_b = connection_manager.clone();
+        let process_b = std::thread::spawn(move || {
+            let connection = manager_b.connection().unwrap();
+            next_number(&connection, &NumberRowType::Stocktake, "store_a").unwrap()
+        });
+
+        let a = process_a.join().unwrap();
+        let b = process_b.join().unwrap();
+        println!("next_number (INSERT) results : a={} b={}", a, b);
+        assert!(a != b);
+
+        let manager_a = connection_manager.clone();
+        let process_a = std::thread::spawn(move || {
+            let connection = manager_a.connection().unwrap();
+            let result: Result<i64, TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    let num = next_number(con, &NumberRowType::Stocktake, "store_a")?;
+                    std::thread::sleep(core::time::Duration::from_millis(TEST_SLEEP_TIME));
+                    Ok(num)
+                });
+            result.unwrap()
+        });
+
+        //Part 2: Concurrent up date both doing updates
```

Comment:
Part 1 should result in an insert and an update. Part 2 should result in 2 updates. Maybe wording should be improved here.

Reply:
I added some debug output and the "insert new row conflict" (second error) is not reached. However, difficult to trigger in a test since you would need to pause in the repository method...